### ISSUE TITLE
Layer Wise Training

### ISSUE THREAD

**Issue by Enverrrr** [ultralytics/yolov5#11081]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello,
I need to train network by split it 'cause I have little free memory. I've tried how to solve it however I couldn't find any solutions. 

Are there any documents I missed out?

**Comment by Enverrrr** [comment]
I'm sorry I couldn't find solution

**Comment by Enverrrr** [comment]
> wave Hello @Enverrrr, thank you for your interest in YOLOv5 rocket! Please visit our star [Tutorials](https://docs.ultralytics.com/yolov5) to get started, where you can find quickstart guides for simple tasks like [Custom Data Training](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) all the way to advanced concepts like [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).
> 
> If this is a bug Bug Report, please provide a **minimum reproducible example** to help us debug it.
> 
> If this is a custom training question Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).
> ## Requirements
> 
> [**Python>=3.7.0**](https://www.python.org/) with all [requirements.txt](https://github.com/ultralytics/yolov5/blob/master/requirements.txt) installed including [**PyTorch>=1.7**](https://pytorch.org/get-started/locally/). To get started:
> 
> ```shell
> git clone https://github.com/ultralytics/yolov5  # clone
> cd yolov5
> pip install -r requirements.txt  # install
> ```
> 
> ## Environments
> 
> YOLOv5 may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
> 
>     * **Notebooks** with free GPU: [![Run on Gradient](https://camo.githubusercontent.com/c9931a1689c37ab786edd3e1e5f59b9a6f7d097628c4689ce2432563ef884524/68747470733a2f2f6173736574732e706170657273706163652e696f2f696d672f6772616469656e742d62616467652e737667)](https://bit.ly/yolov5-paperspace-notebook) [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> 
>     * **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
> 
>     * **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
> 
>     * **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> 
> ## Status
> 
> [![YOLOv5 CI](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml/badge.svg)](https://github.com/ultralytics/yolov5/actions/workflows/ci-testing.yml)
> 
> If this badge is green, all [YOLOv5 GitHub Actions](https://github.com/ultralytics/yolov5/actions) Continuous Integration (CI) tests are currently passing. CI tests verify correct operation of YOLOv5 [training](https://github.com/ultralytics/yolov5/blob/master/train.py), [validation](https://github.com/ultralytics/yolov5/blob/master/val.py), [inference](https://github.com/ultralytics/yolov5/blob/master/detect.py), [export](https://github.com/ultralytics/yolov5/blob/master/export.py) and [benchmarks](https://github.com/ultralytics/yolov5/blob/master/benchmarks.py) on MacOS, Windows, and Ubuntu every 24 hours and on every commit.
> ## Introducing YOLOv8 rocket
> 
> We're excited to announce the launch of our latest state-of-the-art (SOTA) object detection model for 2023 - [YOLOv8](https://github.com/ultralytics/ultralytics) rocket!
> 
> Designed to be fast, accurate, and easy to use, YOLOv8 is an ideal choice for a wide range of object detection, image segmentation and image classification tasks. With YOLOv8, you'll be able to quickly and accurately detect objects in real-time, streamline your workflows, and achieve new levels of accuracy in your projects.
> 
> Check out our [YOLOv8 Docs](https://docs.ultralytics.com/) for details and get started with:
> 
> ```shell
> pip install ultralytics
> ```

I'm sorry I couldn't find solutions

**Comment by Enverrrr** [comment]
@glenn-jocher could you help me please

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about **CUDA memory issues**. YOLOv5 🚀 can be
trained on CPU, [single-GPU](
https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data), or
[multi-GPU](https://github.com/ultralytics/yolov5/issues/475). When
training on GPU it is important to keep your batch-size small enough that
you do not use all of your GPU memory, otherwise you will see a CUDA Out Of
Memory (OOM) Error and your training will crash. You can observe your CUDA
memory utilization using either the `nvidia-smi` command or by viewing your
console output:

<img width="800" alt="Screenshot 2021-05-28 at 12 19 51" src="
https://user-images.githubusercontent.com/26833433/131122318-8141d1ce-13de-44bf-bbdc-487bad442ff0.png
">

### CUDA Out of Memory Solutions

If you encounter a CUDA OOM error, the steps you can take to reduce your
memory usage are:

- Reduce `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s >
YOLOv5n
- Train with [multi-GPU](https://github.com/ultralytics/yolov5/issues/475)
at the same `--batch-size`
- Upgrade your hardware to a larger GPU
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img
src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In
Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="
https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In
Kaggle"></a>

### AutoBatch

You can use [YOLOv5 AutoBatch](
https://github.com/ultralytics/yolov5/pull/5092) (NEW) to find the best
batch size for your training by passing `--batch-size -1`. AutoBatch will
solve for a 90% CUDA memory-utilization batch-size given your training
settings. AutoBatch is experimental, and only works for Single-GPU
training. It may not work on all systems, and is not recommended for
production use.

<img width="1746" alt="Screenshot 2021-11-06 at 12 31 10" src="
https://user-images.githubusercontent.com/26833433/140608327-9c0c0baf-a0ee-4551-a4ff-5ee26b155c51.png
">

Good luck 🍀 and let us know if you have any other questions!

**Comment by glenn-jocher** [comment]
👋 Hello! Thanks for asking about **CUDA memory issues**. YOLOv5 🚀 can be trained on CPU, [single-GPU](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data), or [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training). When training on GPU it is important to keep your batch-size small enough that you do not use all of your GPU memory, otherwise you will see a CUDA Out Of Memory (OOM) Error and your training will crash. You can observe your CUDA memory utilization using either the `nvidia-smi` command or by viewing your console output:

<img width="800" alt="Screenshot 2021-05-28 at 12 19 51" src="https://user-images.githubusercontent.com/26833433/131122318-8141d1ce-13de-44bf-bbdc-487bad442ff0.png">

### CUDA Out of Memory Solutions

If you encounter a CUDA OOM error, the steps you can take to reduce your memory usage are:

- Reduce `--batch-size`
- Reduce `--img-size`
- Reduce model size, i.e. from YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s > YOLOv5n
- Train with [multi-GPU](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training) at the same `--batch-size`
- Upgrade your hardware to a larger GPU
- Train on free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

### AutoBatch

You can use [YOLOv5 AutoBatch](https://github.com/ultralytics/yolov5/pull/5092) (NEW) to find the best batch size for your training by passing `--batch-size -1`. AutoBatch will solve for a 90% CUDA memory-utilization batch-size given your training settings. AutoBatch is experimental, and only works for Single-GPU training. It may not work on all systems, and is not recommended for production use.

<img width="1746" alt="Screenshot 2021-11-06 at 12 31 10" src="https://user-images.githubusercontent.com/26833433/140608327-9c0c0baf-a0ee-4551-a4ff-5ee26b155c51.png">

Good luck 🍀 and let us know if you have any other questions!

**Comment by Enverrrr** [comment]
> 

Thanks for your advice. 
Can I split the network or can I freeze different part of the network. For example train 50 epochs with backbone then 50 epochs head etc.

**Comment by glenn-jocher** [comment]
@Enverrrr that's an unorthodox approach but yes you can use the --freeze command to freeze the first x layers, i.e. `python train.py --freeze 10`

**Comment by Enverrrr** [comment]
> 
I couldn't explain myself I think, I just want to get 1 single weight after training. I mean first of all I need to train backbone and then train other parts of the network with the helps of --freeze command in single run.

**Comment by Enverrrr** [comment]
> @Enverrrr that's an unorthodox approach but yes you can use the --freeze command to freeze the first x layers, i.e. `python train.py --freeze 10`

I couldn't explain myself I think, I just want to get 1 single weight after training. I mean first of all I need to train backbone and then train other parts of the network with the helps of --freeze command in single run.

**Comment by Enverrrr** [comment]
@glenn-jocher Hey I've question for you I don't use freeze method and train all the networks gpu memory is 3142MB according to nvidia-smi. When I am freezing 18 layers from 6 to 24 and gpu memory is 3118MB according to nvidia-smi. Could it be smaller than first situation. I mean yes it is smaller but wouldn't the difference be more?

**Comment by glenn-jocher** [comment]
@Enverrrr Yes, freezing layers can indeed result in a smaller memory footprint, but the extent of the reduction depends on various factors such as the size of the frozen layers and the specific operations performed within those layers. It's not uncommon for the reduction to be less than expected. Additionally, other factors such as the size of the model, batch size, and input image size can also influence GPU memory utilization. If your goal is to minimize memory usage, you may need to further experiment with different combinations of frozen layers and training settings to achieve the desired outcome.
